src/main/frontend/handler/editor.cljs

```diff
@@ -2804,3 +2794,123 @@
     (save-current-block! {:force? true})
     (util/forward-kill-word input)
     (state/set-edit-content! (state/get-edit-input-id) (.-value input))))
+
+(defn tree-seq-with-level
+  [branch? children root]
+  (let [walk (fn walk [level node]
+               (lazy-seq
+                (cons (assoc node :block/level level)
+                      (when (branch? node)
+                        (mapcat (partial walk (inc level)) (children node))))))]
+    (walk 1 root)))
+
+(defn all-blocks-with-level
+  "Return all blocks associated with correct level
+   if :collapse? true, return without any collapsed children
+   for example:
+   - a
+    - b (collapsed)
+     - c
+     - d
+    - e
+   return:
+    blocks
+    [{:block a :level 1}
+     {:block b :level 2}
+     {:block e :level 2}]"
+  [{:keys [collapse?] :or {collapse? false}}]
+  (let [page (state/get-current-page)]
+    (cond->>
+     (-> page
+         (db/get-page-blocks-no-cache)
+         (tree/blocks->vec-tree page))
+
+     collapse?
+     (w/postwalk
+      (fn [x]
+        (if (and (map? x) (-> x :block/properties :collapsed))
+          (assoc x :block/children [])
+          x)))
+
+     :default
+     (mapcat (fn [x] (tree-seq-with-level map? :block/children x)))
```

Comment:
This commit https://github.com/logseq/logseq/commit/f71f32475a7183bc97844bf4bec138c5c91782c7 added `:block/level` to each block returned by `(tree/blocks->vec-tree root-block)`.

Reply:
fixed